File: dustroute/port_realization.py
from __future__ import annotations

from dataclasses import dataclass

from .cells import PortKind
from .model import Facing, Pos, World


_HORIZONTAL = {
    Facing.NORTH:(0,0,-1),
    Facing.EAST:(1,0,0),
    Facing.SOUTH:(0,0,1),
    Facing.WEST:(-1,0,0),
}


@dataclass(frozen=True)
class PortRealization:
    """Physical routing contract derived from one typed endpoint."""
    terminal: Pos
    approach: Pos
    leaf_required: bool
    approach_facing: Facing | None


def terminal_for_endpoint(world: World, endpoint) -> Pos:
    """Translate a typed port into the conductor position the router targets."""
    if endpoint.kind is PortKind.WIRE:
        return endpoint.pos
    if endpoint.kind is PortKind.BLOCK_POWER:
        delta=_HORIZONTAL.get(endpoint.facing)
        if delta is None:
            raise ValueError("BLOCK_POWER port requires horizontal facing")
        return endpoint.pos.offset(*delta)
    raise ValueError(endpoint.kind)


def realize_sink_endpoint(world: World, endpoint) -> PortRealization:
    terminal=terminal_for_endpoint(world,endpoint)
    delta=_HORIZONTAL.get(endpoint.facing)
    if delta is None:
        return PortRealization(
            terminal=terminal,
            approach=terminal,
            leaf_required=True,
            approach_facing=None,
        )
    return PortRealization(
        terminal=terminal,
        approach=terminal.offset(*delta),
        leaf_required=True,
        approach_facing=endpoint.facing,
    )


def realize_source_endpoint(world: World, endpoint) -> PortRealization:
    terminal=terminal_for_endpoint(world,endpoint)
    # Sources may fan out after their terminal. A directional output can later
    # gain a fixed departure policy without changing the router API.
    return PortRealization(
        terminal=terminal,
        approach=terminal,
        leaf_required=False,
        approach_facing=endpoint.facing,
    )

File: crates/dustroute-translate/src/port_realization.rs
use std::error::Error;
use std::fmt::{Display, Formatter};

use crate::cells::PortKind;
use crate::physical::Endpoint;
use crate::world::{Facing, Pos};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct PortRealization {
    pub terminal: Pos,
    pub approach: Pos,
    pub leaf_required: bool,
    pub approach_facing: Option<Facing>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PortRealizationError {
    BlockPowerRequiresHorizontalFacing,
}

impl Display for PortRealizationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("BLOCK_POWER port requires horizontal facing")
    }
}

impl Error for PortRealizationError {}

pub fn terminal_for_endpoint(endpoint: &Endpoint) -> Result<Pos, PortRealizationError> {
    match endpoint.kind {
        PortKind::Wire => Ok(endpoint.pos),
        PortKind::BlockPower => {
            let delta = endpoint
                .facing
                .and_then(Facing::horizontal_offset)
                .ok_or(PortRealizationError::BlockPowerRequiresHorizontalFacing)?;
            Ok(endpoint.pos.offset(delta.x, delta.y, delta.z))
        }
    }
}

pub fn realize_sink_endpoint(endpoint: &Endpoint) -> Result<PortRealization, PortRealizationError> {
    let terminal = terminal_for_endpoint(endpoint)?;
    let Some(delta) = endpoint.facing.and_then(Facing::horizontal_offset) else {
        return Ok(PortRealization {
            terminal,
            approach: terminal,
            leaf_required: true,
            approach_facing: None,
        });
    };
    Ok(PortRealization {
        terminal,
        approach: terminal.offset(delta.x, delta.y, delta.z),
        leaf_required: true,
        approach_facing: endpoint.facing,
    })
}

pub fn realize_source_endpoint(
    endpoint: &Endpoint,
) -> Result<PortRealization, PortRealizationError> {
    let terminal = terminal_for_endpoint(endpoint)?;
    Ok(PortRealization {
        terminal,
        approach: terminal,
        leaf_required: false,
        approach_facing: endpoint.facing,
    })
}

#[cfg(test)]
mod tests {
    use crate::physical::PlacementCircuit;

    use super::*;

    #[test]
    fn block_power_targets_external_conductor() {
        let endpoint = PlacementCircuit::boundary(
            "a",
            Pos::new(10, 2, 4),
            PortKind::BlockPower,
            Some(Facing::West),
        );
        let realized = realize_sink_endpoint(&endpoint).unwrap();
        assert_eq!(realized.terminal, Pos::new(9, 2, 4));
        assert_eq!(realized.approach, Pos::new(8, 2, 4));
        assert!(realized.leaf_required);
    }
}
